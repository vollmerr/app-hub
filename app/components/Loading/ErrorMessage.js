import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';

import Link from 'components/Link';

import Wrapper from './Wrapper';

const Header = styled.div`
  font-size: 24px;
`;

const Message = styled.div`
  font-size: 14px;
`;

const Button = styled(DefaultButton) `
  margin: 20px;
`;

class ErrorMessage extends React.PureComponent {
  render() {
    const { error } = this.props;
    const message = error && error.message;

    return (
      <Wrapper>
        <Header>Sorry, an error has occurred.</Header>
        {message && <Message>{message}</Message>}
        <Link to={'/'}><Button primary>Home</Button></Link>
      </Wrapper>
    );
  }
}

ErrorMessage.propTypes = {
  error: PropTypes.object,
};

export default ErrorMessage;
