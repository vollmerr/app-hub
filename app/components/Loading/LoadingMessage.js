import React from 'react';
import PropTypes from 'prop-types';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';

import Wrapper from './Wrapper';


class LoadingMessage extends React.PureComponent {
  render() {
    const { text } = this.props;

    return (
      <Wrapper>
        <Spinner size={SpinnerSize.large} label={text || 'Loading...'} ariaLive={'assertive'} />
      </Wrapper>
    );
  }
}


const { string } = PropTypes;

LoadingMessage.propTypes = {
  text: string,
};

export default LoadingMessage;
