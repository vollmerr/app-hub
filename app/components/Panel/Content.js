import styled from 'styled-components';

import theme from 'utils/theme';

const Content = styled.div`
  background: ${theme.neutralSecondary};
  color: ${theme.white};
  font-size: 16px;
  bottom: 0;
  ${(props) => props.left ? 'left' : 'right'}: 0;
  position: absolute;
  width: ${theme.hub.panelWidth};
  height: calc(100vh - ${theme.hub.headerHeight});
  border-${(props) => props.left ? 'right' : 'left'}: 1px solid ${theme.neutralSecondary};
  box-sizing: border-box;
  cursor: default;
  z-index: ${theme.zIndex.panel};
`;

export default Content;
