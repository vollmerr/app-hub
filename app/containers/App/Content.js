
import styled from 'styled-components';

import theme from 'utils/theme';

const Content = styled.div`
  display: flex;
  flex: 1;
  padding: 0 ${(props) => props.isMobile ? '15px' : '30px'};
  background-color: ${theme.neutralLighterAlt};
`;

export default Content;
