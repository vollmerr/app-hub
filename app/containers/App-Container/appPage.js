import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { history } from 'configureStore';
import ErrorMessage from 'components/Loading/ErrorMessage';
import LoadingMessage from 'components/Loading/LoadingMessage';
import Content from 'components/App-Content/Content';
import { makeSelectApp, makeSelectUserRoles } from 'containers/AppHub/selectors';


export function appPage(Component) {
  return class extends React.PureComponent {
    static displayName = Component.displayName ||
      Component.name /* istanbul ignore next: can't set `name` for testing as read only */ ||
      'AppPage';

    static propTypes = {
      app: PropTypes.object.isRequired,
      userRoles: PropTypes.object.isRequired,
    };

    componentWillReceiveProps(nextProps) {
      const { app, userRoles } = this.props;
      // if there are different routes (is empty when component loads, so check here)
      if (nextProps.app.get('routes') !== app.get('routes')) {
        // get index of current route
        const currentRoute = nextProps.app
          .get('routes')
          .find((route) => (
            route.get('path') === history.location.pathname)
          );
        // route found, that has permissions set
        if (currentRoute && currentRoute.get('roles')) {
          // not authorized for route
          if (!currentRoute.get('roles').some((role) => userRoles.includes(role))) {
            // get home route for app
            const homeRoute = nextProps.app
              .get('routes')
              .find((route) => (
                route.get('key').match(/Home/))
              );
            // push to home of app
            history.push(homeRoute ? homeRoute.get('path') : '/');
            // dont update
            return false;
          }
        }
      }
      return nextProps !== this.props;
    }

    render() {
      const { app } = this.props;
      const routes = app.get('routes');
      const homeRoute = routes.find((route) => (
        route.get('key').match(/Home/))
      );
      const to = homeRoute ? homeRoute.get('path') : '/';

      if (app.get('error')) {
        return <ErrorMessage error={app.get('error')} to={to} />;
      }

      let Loading = null;
      if (app.get('loading')) {
        Loading = <LoadingMessage />;
      }

      return (
        <Content>
          <Component {...this.props} Loading={Loading} />
        </Content>
      );
    }
  };
}

const mapStateToProps = createStructuredSelector({
  app: makeSelectApp(),
  userRoles: makeSelectUserRoles(),
});

export default (Component) => connect(mapStateToProps, {})(appPage(Component));
