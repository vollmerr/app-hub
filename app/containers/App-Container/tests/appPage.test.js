import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import ErrorMessage from 'components/Loading/ErrorMessage';
import LoadingMessage from 'components/Loading/LoadingMessage';
import Content from 'components/App-Content/Content';
import { history } from 'configureStore';

import { appPage } from '../appPage';

const Component = () => <div>test component</div>;

const userRoles = ['test role1', 'test role 2'];
const routes = [{ key: 'testHome', path: '/test' }, { key: 'key2', path: '/test2' }];

const props = {
  app: fromJS({
    error: null,
    loading: false,
    routes,
  }),
  userRoles: fromJS(userRoles),
};

history.push = jest.fn();

const error = 'test error';

describe('appPage', () => {
  let wrapper;
  let WithAppPage;
  let instance;
  let newProps;
  beforeEach(() => {
    jest.resetAllMocks();
    WithAppPage = appPage(Component);
    wrapper = shallow(<WithAppPage {...props} />);
    instance = wrapper.instance();
    newProps = { ...props };
  });

  it('should be a HOC that returns a new component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Component passed', () => {
    expect(wrapper.find(Component).length).toEqual(1);
  });

  it('should adopt the correct displayName', () => {
    // using `displayName`
    Component.displayName = 'testDisplayName';
    WithAppPage = appPage(Component);
    expect(WithAppPage.displayName).toEqual(Component.displayName);
    // using `name` from function style component
    Component.displayName = undefined;
    WithAppPage = appPage(Component);
    expect(WithAppPage.displayName).toEqual(Component.name);
  });

  it('should wrap the rendered Component in `Content` for styling', () => {
    expect(wrapper.find(Content).find(Component).length).toEqual(1);
  });

  it('should render an error message if passed an error', () => {
    newProps.app = newProps.app.set('error', error);
    wrapper.setProps({ ...newProps });
    expect(wrapper.find(ErrorMessage).length).toEqual(1);
  });

  it('should pass the `error` and `to` path to the error message', () => {
    newProps.app = newProps.app.set('error', error);
    wrapper.setProps({ ...newProps });
    expect(wrapper.find(ErrorMessage).prop('error')).toEqual(error);
    expect(wrapper.find(ErrorMessage).prop('to')).toEqual(routes[0].path);
  });

  it('should pass the AppHub base route to the error message if no `to` provided', () => {
    newProps.app.routes = fromJS([]);
    newProps.app = newProps.app.set('error', error);
    newProps.app = newProps.app.set('routes', fromJS([]));
    wrapper.setProps({ ...newProps });
    expect(wrapper.find(ErrorMessage).prop('to')).toEqual('/');
  });

  it('should pass a loading component if loading', () => {
    newProps.app = newProps.app.set('loading', true);
    wrapper.setProps({ ...newProps });
    expect(wrapper.find(Component).prop('Loading')).toEqual(<LoadingMessage />);
  });


  describe('componentWillReceiveProps', () => {
    it('should update as normal when non route props change', () => {
      newProps.app = newProps.app.set('loading', true);
      expect(instance.componentWillReceiveProps(newProps)).toEqual(true);
      expect(history.push).not.toHaveBeenCalled();
    });

    it('should update as normal if going to a valid route with no permissions needed', () => {
      newProps.app = newProps.app.set('routes', fromJS([{ key: 'key3', path: '/test3' }]));
      history.location.pathname = '/test3';
      expect(instance.componentWillReceiveProps(newProps)).toEqual(true);
      expect(history.push).not.toHaveBeenCalled();
    });

    it('should update as normal if going to a valid route user has permissions for', () => {
      newProps.app = newProps.app.set('routes', fromJS([{ key: 'key3', path: '/test3', roles: [userRoles[0]] }]));
      history.location.pathname = '/test3';
      expect(instance.componentWillReceiveProps(newProps)).toEqual(true);
      expect(history.push).not.toHaveBeenCalled();
    });

    it('should redirect to the app home if going to a route user doesnt have permissions for', () => {
      newProps.app = newProps.app.set('routes', fromJS([routes[0], { key: 'key3', path: '/test3', roles: ['otherRole'] }]));
      history.location.pathname = '/test3';
      expect(instance.componentWillReceiveProps(newProps)).toEqual(false);
      expect(history.push).toHaveBeenCalledWith(routes[0].path);
    });

    it('should redirect to the AppHub home if going to a route user doesnt have permissions for and no app home route', () => {
      newProps.app = newProps.app.set('routes', fromJS([{ key: 'key3', path: '/test3', roles: ['otherRole'] }]));
      history.location.pathname = '/test3';
      expect(instance.componentWillReceiveProps(newProps)).toEqual(false);
      expect(history.push).toHaveBeenCalledWith('/');
    });
  });
});

