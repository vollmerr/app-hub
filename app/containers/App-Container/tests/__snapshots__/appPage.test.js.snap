// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appPage should be a HOC that returns a new component 1`] = `
<Content__Content>
  <Component
    Loading={null}
    app={
      Immutable.Map {
        "error": null,
        "loading": false,
        "routes": Immutable.List [
          Immutable.Map {
            "key": "testHome",
            "path": "/test",
          },
          Immutable.Map {
            "key": "key2",
            "path": "/test2",
          },
        ],
      }
    }
    userRoles={
      Immutable.List [
        "test role1",
        "test role 2",
      ]
    }
  />
</Content__Content>
`;
